lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {DeepRecord, Key, CollectionKey, Selector, Value} from './types';
+
+type KeyValueMap = {
+    [TKey in Key | CollectionKey]: Value[TKey];
+};
+
+type BaseConnectOptions<TKey extends Key | CollectionKey> = {
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Selector<TKey, Value[TKey] | null>;
+};
+
+type ConnectOptions<TKey extends Key | CollectionKey> = BaseConnectOptions<TKey> &
+    (
+        | {
+              key: TKey extends CollectionKey ? TKey : never;
+              callback?: (value: Record<string, Value[TKey]> | null, key?: TKey) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              key: TKey;
+              callback?: (value: Value[TKey] | null, key?: TKey) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+type MergeCollection<TKey extends Key, TMap, TValue> = {
+    [MapK in keyof TMap]: MapK extends `${TKey}${string}`
+        ? MapK extends `${TKey}`
+            ? never // forbids empty id
+            : TValue
+        : never;
+};
+
+type UpdateOperation<TKey extends Key | CollectionKey> =
```

Comment:
To match the naming on the backend it would be nice to call this type `OnyxUpdate`